### PR TITLE
Go back to capturing *non-user* modules by 'require' reference.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes: https://github.com/pulumi/pulumi/issues/1588

This PR brings module serialization closer to what it used to be.  Specifically, if we see a module captured in a serialized function we will serialize it out into the function as ```require(<module_name>)```.  This means that it is now possible once more to just ```import * as fs from "fs"``` at the top level of a pulumi program and use it within a serialized callback function.

However, there are some slight tweaks to the previous approach that we've taken here that address several of the shortcoming of the previous approach.

1.  Modules can optionally opt-into stating that they do not want to be captured in this manner, and that they should instead be captured by the current 'by value' serialization semantics we have elsewhere.  An example of a module that does this is our 'Config' module.  By capturing as a value (and not as a `require`), the Config object will hydrate on the cloud-callback side with all the configuration values specified at deployment time.
2. We capture the *local* module (i.e. the user's code) by value.  This prevents references to local code from breaking like it used to.  This is necessary so that user code does not try to actually "require" other user code (which would not work, as none of the user code files are actually there at runtime).  Instead, because the local module is entirely captured by value, all references to other user code naturally gets slurped into the closure serialization and is directly available in the cloud lambda.

With this code can revert to the normal idiomatic TS/JS form with all 'imports' at the top level.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@lukehoban I've tested this downstream and things seem to be working well.  Can you take a look so we can get this in early and get time to bang on this?

**Comment by CyrusNajmabadi** [comment]
@lukehoban this now has the changes we discussed before you left.  'Config' is now strictly an object that is properly capturable.  

Note: this did take some finagling.  Specifically, to make it capturable, i removed all dependencies on things that prevented that.  For example, it can no longer use "log.warn" (as that ends up calling into the pulumi-engine), but instead just does a "console.log".  This will have no observable change in behavior for the user as these warnings still stay parented by the stack.

**Comment by CyrusNajmabadi** [comment]
@lukehoban Can you review?   I'd like to get this in asap.  Thanks!

File: sdk/nodejs/config.ts [hunk 1]
```diff
@@ -34,11 +33,10 @@ export class Config {
         // To ease migration of code that already does new Config("<package>:config"), treat this as if
         // just new Config("<package>") was called.
         if (name.endsWith(":config")) {
-            const newName = name.replace(/:config$/, "");
-            log.warn(util.format("`:config` is no longer required at the end of configuration " +
+            name = name.replace(/:config$/, "");
+            console.log(util.format("`:config` is no longer required at the end of configuration " +
```
Comment (swgillespie) [inline sdk/nodejs/config.ts]: I assume this is because `log.warn` can't be closure serialized?

File: sdk/nodejs/runtime/closure/createClosure.ts [hunk 2]
```diff
@@ -749,23 +752,56 @@ function getOrCreateEntry(
     dispatchAny();
     return entry;
 
-    function dispatchAny(): void {
+    function doNotCapture() {
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply sdk/nodejs/runtime/closure/createClosure.ts]: i'm fine with inferred return types.  but i can add this if you would prefer it.

File: sdk/nodejs/runtime/closure/createClosure.ts [hunk 3]
```diff
@@ -749,23 +752,56 @@ function getOrCreateEntry(
...
             typeof obj === "boolean" || typeof obj === "number" || typeof obj === "string") {
             // Serialize primitives as-is.
             entry.json = obj;
+            return;
+        }
+
+        const moduleName = findModuleName(obj);
+        if (moduleName && moduleName[0] !== ".") {
+            if (moduleName.startsWith("@pulumi")) {
+                throw new RunError("'@pulumi' modules cannot be used inside a cloud-callback.");
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply sdk/nodejs/runtime/closure/createClosure.ts]: You're correct.  This needs to be throwSerializationError.  Thanks!

File: sdk/nodejs/runtime/closure/createClosure.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (swgillespie) [reply sdk/nodejs/runtime/closure/createClosure.ts]: up to you!

File: sdk/nodejs/tests/runtime/closure.spec.ts [hunk 4]
```diff
@@ -4699,6 +4770,217 @@ return function () { console.log(o1); console.log(o2.b.d); console.log(o3.b.d);
...
+  }).apply(undefined, undefined).apply(this, arguments);
+}
+
+function __ensureConfig() {
+  return (function() {
+    with({ loaded: true, config: __config, cleanKey: __cleanKey, ensureConfig: __ensureConfig }) {
+
+return function /*ensureConfig*/() {
+    if (!loaded) {
+        const envConfig = process.env.PULUMI_CONFIG;
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply sdk/nodejs/tests/runtime/closure.spec.ts]: It 'works' in that *if* you were to have this in your environment, we would read out the values of it.  But, that only happens if the user sets this value.

Note: the reason Config works is actually because we serialize the __config global over.

File: sdk/nodejs/runtime/closure/createClosure.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply sdk/nodejs/runtime/closure/createClosure.ts]: I'll leave as is.  Thanks! :)

File: sdk/nodejs/runtime/closure/createClosure.ts [hunk 5]
```diff
@@ -748,23 +751,74 @@ function getOrCreateEntry(
...
+                // creation is because we're serializing out some function.
+                let func: Function = <any>undefined;
+                for (let i = context.frames.length - 1; i >= 0; i--) {
+                    const frame = context.frames[i];
+                    if (frame.functionLocation) {
+                        func = frame.functionLocation.func;
+                    }
+                }
+
+                throwSerializationError(func, context,
```
Comment (joeduffy) [inline sdk/nodejs/runtime/closure/createClosure.ts]: This seems semi-arbitrary to me; can you remind me why we need to do this? Firstly, virtually anything can be put into a @pulumi module, so it's not clear that by definition eliding them won't lead someone to be in a situation where they can't access important functionality (like a helper function). Secondly, hard coded rules like this always frustrate me as a user.

If it's just that gRPC is too large, why don't we filter out native modules, or something more generalizable and principled
?

Comment (CyrusNajmabadi) [reply sdk/nodejs/runtime/closure/createClosure.ts]: > This seems semi-arbitrary to me; can you remind me why we need to do this?

For one, because we explicitly do not upload any @pulumi node_modules.   So if we actually capture a pulumi-module *by* reference, then that will def break at runtime. :)

This is both because that node_module is enormous, and because much of it does not work on the 'inside' due to needing to be able to do things like speak to the pulumi engine.

> so it's not clear that by definition eliding them won't lead someone to be in a situation where they can't access important functionality (like a helper function). 

Note: using a helper function is fine.  In essence, that's what is happening when you use "pulumi.Config".  If you actually use anything *in* the module, then that is captured by value 'as expected', and it serialized properly into your lambda code.  

What is being prevented here is literally being able to actually use the module-object itself.  

> If it's just that gRPC is too large, why don't we filter out native modules, 

I'm not sure what counts as a native module :)   We definitely don't want to prevent a module that *uses* anything native.  in practice, most modules will end up doing that in some manner.  

> or something more generalizable and principled?

I think the problem is that it's really hard to find an actual 'principle' that doesn't have some problem.  For example, our old principle was "do not special case modules.  capture everything by value."  That was principled, but broke down badly with modules (forcing people to do things like 'require on the inside').  Similarly, we tried the principle of capturing all modules by by 'import'.  But this ends up breaking user-code itself because now we don't actually properly serialize code that can work properly since our serialization strategy does not properly mimic the user's actual code layout (and trying to do so would be a large undertaking).

This approach tries to accomplish the following:

1. allow users to write the code in idiomatic TS/JS, and have the semantics for all their module code behave as expected.
2. have a capturing story that has simple and explainable semantics.
3. allow people to author  a single project.json that defines both the deployment-time and cloud-runtime-time configuration, with reasonable outputs for both environments.

After this PR we have the above.  However, there is one specific issue a user could potentially trip themselves into.  Namely, they end up wanting to actually capture a pulumi module reference as a reference.  

Right now we can't think of any good reason why someone would want to do this.  And, instead of having it silently blow up, i thought it would be good to at least have us let the user know they're doing something bad.  

Note: In the future, if we think there is a good reason for this.  We can definitely relax this restriction.  But i'd rather start with a good working system, with a single non-onerous restriction, and only relax later if we really think it's necessary.

Comment (CyrusNajmabadi) [reply sdk/nodejs/runtime/closure/createClosure.ts]: @lukehoban Would def appreciate your thoughts here.

Comment (joeduffy) [reply sdk/nodejs/runtime/closure/createClosure.ts]: > This is both because that node_module is enormous

But, why is it enormous? And why is an enormous package specific in any way to `@pulumi`?

> much of it does not work on the 'inside' due to needing to be able to do things like speak to the pulumi engine.

Yes, but the key word in here is _much_. There is some of it that could work.

I understand that the pragmatic, expedient thing here is to just filter out `@pulumi`, but it really doesn't feel like this is where we want to land long-term.

This PR is a great step forward, so I'm not blocking this change, but any magic name-based special casing is always a big code smell to me, and this part has me a little concerned.

**Comment by CyrusNajmabadi** [comment]
Preliminary docs on this subject are here: https://github.com/pulumi/docs/pull/501

File: sdk/nodejs/config.ts [hunk 6]
```diff
@@ -34,11 +33,10 @@ export class Config {
         // To ease migration of code that already does new Config("<package>:config"), treat this as if
         // just new Config("<package>") was called.
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply sdk/nodejs/config.ts]: Done.

File: sdk/nodejs/runtime/closure/createClosure.ts (hunk 5 above)
Comment (CyrusNajmabadi) [reply sdk/nodejs/runtime/closure/createClosure.ts]: > But, why is it enormous? 

The primary reasons it is enormous is because pulumi pulls in both grpc (56MB) and typescript (35mb).  Together, we're already close to 100MB on just these two packages.  These two packages are only needed by pulumi at deployment time to:

1. Use GRPC to communicate with teh pulumi engine (something not needed at cloud-runtime). 
2. Use TypeScript ot analyzer serialize a javascript function into the code needed by the AWS lambda (something not needed at cloud-runtime).

> And why is an enormous package specific in any way to @pulumi?

It's specific to pulumi because the @pulumi libraires are the ones that are so specifically intended primarily for deployment time. This is different form (literally) every other library out there, which very much *is* intended to be used at runtime by an application. 

It's also specific to @pulumi in that users have no control here about including those packages.  You have to include those packages, or else pulumi simply doesn't work.  However, any other packages you don't need, you can simply leave out of your package.json.  However, Pulumi Applications are themselves special in that they have two very distinct lifetimes.  The local deployment lifetime, and the remote cloud lifetime.  @pulumi is needed for the former, but we don't want the usage int he former to cost you for the latter. :)

> Yes, but the key word in here is much. There is some of it that could work.

Note: it's important to be very precise about what does or doesn't work here.  If you end up trying to use part of pulumi which *can* run in the cloud then it *will* work.  That's part of this work.  The poster-child for this is "pulumi.Config".  You can use Config inside the cloud because it is specifically accessing the part of the pulumi that will work, while side stepping the parts that will not work. 

Similarly, you can obviously capture and use any sort of pulumi resource (even though those came from an @pulumi module).  What does *not* work is trying to write code that effectively would become a `require("@pulumi")` at runtime.  So far, there are no scenarios we can envision where could would want to actually do that.

Comment (CyrusNajmabadi) [reply sdk/nodejs/runtime/closure/createClosure.ts]: > This PR is a great step forward, so I'm not blocking this change, but any magic name-based special casing is always a big code smell to me, and this part has me a little concerned.

Sounds good to me.  Note that IMO, it's much easier to start with carving out things that we do not want to support up front.  Later on, we can relax in a conscientious and deliberate fashion based on very intentional and specific scenarios to support.

Comment (lukehoban) [reply sdk/nodejs/runtime/closure/createClosure.ts]: I do think it would be good to open the issue tracking adding a first class marker for this instead of relying on `@pulumi/` scope.  And then link that issue as `TODO` here in the code where we hard code `@pulumi/`.

File: sdk/nodejs/runtime/closure/createClosure.ts [hunk 7]
```diff
@@ -748,23 +751,74 @@ function getOrCreateEntry(
...
+
+        if (obj === undefined || obj === null ||
             typeof obj === "boolean" || typeof obj === "number" || typeof obj === "string") {
             // Serialize primitives as-is.
             entry.json = obj;
+            return;
+        }
+
+        const moduleName = findModuleName(obj);
+        if (moduleName && moduleName[0] !== ".") {
```
Comment (lukehoban) [inline sdk/nodejs/runtime/closure/createClosure.ts]: The logic here is a little hard to follow.  I think there are three cases:
1. A local module
2. A non-local module marked with the "no serialize" bit (currently hardcoded to `@pulumi` scope)
3. Everything else

Can the logic here explicitly branch to those three cases, and call out what the precondition is and why each behaves the way it does?

File: sdk/nodejs/runtime/closure/createClosure.ts [hunk 8]
```diff
@@ -748,23 +751,74 @@ function getOrCreateEntry(
...
+
+                throwSerializationError(func, context,
+                    "'@pulumi' modules cannot be referenced inside a function that will execute in the Cloud.");
+            }
+
+            // This name bound to a module, and the module did not opt into having itself captured
+            // by value. In this case serialize it out as a direct 'require' call
+            //
+            // Also, importantly, if this is a reference to a local module (i.e. starts with '.'),
+            // then always capture it as a value.  The reason for this is that otherwise we won't
```
Comment (lukehoban) [inline sdk/nodejs/runtime/closure/createClosure.ts]: Is it correct to treat local modules differently from `@pulumi` modules?  What's the rationale? 

* Local: capture by value
* `@pulumi`: reject

Should these both reject, or both capture by value?  The prose here doesn't make clear (at least to me) why these are different.

File: sdk/nodejs/runtime/closure/createClosure.ts (hunk 7 above)
Comment (CyrusNajmabadi) [reply sdk/nodejs/runtime/closure/createClosure.ts]: Sure.

File: sdk/nodejs/runtime/closure/createClosure.ts (hunk 8 above)
Comment (CyrusNajmabadi) [reply sdk/nodejs/runtime/closure/createClosure.ts]: > Should these both reject, or both capture by value? The prose here doesn't make clear (at least to me) why these are different.

I'll improve the prose.  However, basically, both of these are ones we cannot capture by reference.  We could just default to capturing both by value.  

**However**: it's known ahead of time that capturing @pulumi by value simply will not work (simply because it itself references packages that we won't be uploading).  We could just allow the capturing to seem to succeed (or potentially fail), but the user will most likely then get a cryptic error later on down the line.

The point of erroring early is simply so we can provide good failure path.  In other words, using @pulumi is just going ot fail at this part of our product cycle.  We can either make it fail very cryptically (i.e. errors at runtime, or with suggestions to change how it is imported), or we can detect up front this bad pattern and give a *very good* message about what is going wrong.  

Normally, i'm really not a fan of special casing, or having magic strings.  And i definitely be up for some sort of cleaner mechanism to address this issue.  However, i think this is a very low cost, and *effective* mechanism for quick and clearly helping people avoid something that is almost surely going to otehrwise be a bad experience :)

--

Note: i'm ok with opening a bug to track investigating other things here.  However, i have a real concern that such bugs simply sit around not serving much purpose since they'll almost always (in isolation) be a low priority for us.  I'd rather us just accept this as an acceptable starting point, and we can drive any changes here through new scenarios with clear customers (internal or external) that need those scenarios to work.

File: sdk/nodejs/runtime/closure/createClosure.ts (hunk 7 above)
Comment (CyrusNajmabadi) [reply sdk/nodejs/runtime/closure/createClosure.ts]: I've extracted a function for handling modules, and i've tried to clearly break out and doucment these cases.

**Comment by CyrusNajmabadi** [comment]
I'm merging this in.  I think we're happy with the overall direction (including have user docs for this area).  If there are remaining concerns, i think it would be fine to handle them in followup PRs.  But i want to get this in asap so we have time to properly use this.

File: sdk/nodejs/runtime/closure/createClosure.ts (hunk 5 above)
Comment (joeduffy) [reply sdk/nodejs/runtime/closure/createClosure.ts]: Having a `@deploymentOnly()` marker would make me feel better. The reason I grow concerned when I see `@pulumi` hard-coded is (1) there are potentially libraries that would be available in `@pulumi` packages that are _not_ just deployment-time, and (2) there are potentially libraries outside of `@pulumi` (such as 3rd party ones) that _are_ only deployment-time.

Note that if we only did this for `@pulumi/pulumi`, our runtime SDK, I'd not have felt the same way.

Comment (CyrusNajmabadi) [reply sdk/nodejs/runtime/closure/createClosure.ts]: https://github.com/pulumi/pulumi/issues/1676 Tracks this.  Note: my preference would be that work here be driven by a real use case that we can use to vet any solution we create.
